src/aff_expr.jl

```diff
@@ -16,12 +16,13 @@
 #############################################################################
 
 # Utilities for OrderedDict
-function _add_or_set!(dict::OrderedDict{K,V}, k::K, v::V) where {K,V}
-    # TODO: This unnecessarily requires two lookups for k.
+function _add_or_set!(exprterms::OrderedDict{K,V}, k::K, v::V) where {K,V}
     # TODO: Decide if we want to drop zeros here after understanding the
     # performance implications.
-    dict[k] = get!(dict, k, zero(V)) + v
-    return dict
+    v0 = get!(exprterms, k, v)
+    # This short circuit makes sure only one hash lookup is need when inserting a value
+    v0 == v || (exprterms[k] = v0 + v)
```

Comment:
IMO the following is easier to read than the short circuit:
```julia
if v0 !== v
    expr_terms[k] = v0 + v
end
```
(`===`/`!==` can be slightly faster than `==`/`!=` in this case, and is safe to use because the documentation for `get!` says that `v` itself will be returned if there's no key defined.)